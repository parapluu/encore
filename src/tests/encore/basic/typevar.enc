local trait AB[a, b]
  require var aa : a
  require var bb : b
  def fn(aa : a) : b
    this.aa = aa
    this.bb
  end
end
class Foo[a, b] : AB[a, b]
  var aa : a
  var bb : b
  def init(aa : a, bb : b) : unit
    this.aa = aa
    this.bb = bb
  end
end
local class Bar[b, c]
  var foo : AB[b, c]
  def init(bb : b, cc : c) : unit
    this.foo = new Foo[b, c](bb, cc)
  end
  def fn(bb : b) : c
    this.foo.fn(bb)
  end
end
active class Main
  def main() : unit
    let
      bar = new Bar[real, int](3.14, 42)
    in
      println("assert(bar.fn(4.13) == 42): {}", bar.fn(4.13))
    end
  end
end
