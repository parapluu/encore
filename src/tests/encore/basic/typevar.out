assert(bar.fun(4.13) == 42): 42
